Регистрация пользователей

AveGus opened the pull request:
*Изменена модель пользователя
*Добавлена бизнес-логика регистрации и получение апи-ключа пользователя(сырая версия) 
*Добавлены pydantic модели
*Добавлен новый api router для регистрации пользователя

AveGus committed:
*Изменена модель пользователя
*Добавлена бизнес-логика регистрации и получение апи-ключа пользователя(сырая версия)
*Добавлены pydantic модели
*Добавлен новый api router для регистрации пользователя

SiMiZZZ commented on src/services/types.py:
```diff
@@ -0,0 +1,6 @@
+import enum
+
+
+class UserRole(enum.Enum):
```
strEnum лучше, нагляднее
ну и лучше сразу импортировать конечный класс вместо импорта enum.Enum

SiMiZZZ commented on src/services/users.py:
```diff
@@ -0,0 +1,23 @@
+
+class CreatedUser(BaseModel):
+    name: str
+    role: str
```
тайпхинт UserRole

SiMiZZZ commented on src/services/users.py:
```diff
@@ -0,0 +1,23 @@
+
+class UserCreate(BaseModel):
+    name: str
+    role: UserRole = UserRole.USER
```
Значение по дефолту, если это не None делать не самая лучшая идея
В момент создания этой модели пропадает явность и читаемось, кажется что в UserCreate передается только name, хотя это не так

Лучше руками передавать все параметры

SiMiZZZ commented on src/services/users.py:
```diff
@@ -0,0 +1,23 @@
+    api_key: str
+
+
+async def get_current_user(api_key: str = Header(...)):
```
спорный нейминг
что такое current в этом контексте - не до конца очевидно

А также нужно пояснение про синтаксит Header(...)
Что тут происходит?

SiMiZZZ replied:
Пояснение в плане мне
Пока выглядит как магия

SiMiZZZ replied:
Нет тайпхинтов возвоащаемого значения

AveGus replied:
> спорный нейминг что такое current в этом контексте - не до конца очевидно
> 
> А также нужно пояснение про синтаксит Header(...) Что тут происходит?

Ну смотри у нас в апи-схеме написано что апи кей будет передаваться в заголовках в формате Authorization: TOKEN key-bee6de4d-7a23-4bb1-a048-523c2ef0ea0c
Я нагуглил что так фаст-апи умеет ловить то что в заголовке(но я щас понял что то что у меня написано чутка неправильно и он не то ловит() передается и вщ эта функция у меня была написана для depends, чтобы в функции можно было отдавать user: User = Depends(get_current_user)

SiMiZZZ commented on src/services/auth/login.py:
```diff
@@ -0,0 +1,16 @@
+from services.users import UserCreate
+
+
+async def register(user_data: UserCreate):
```
неочевидный нейминг

что такое register в этом контексте?
регистрация чего?

+нет тайпхинтов возвращаемого значения(странно, что майпай не дал по жопе)

SiMiZZZ replied:
В функции используется только name и role
Не до конца понятно зачем передавать всю пайдантик модель
Если захочешь зарегать пользователя где-то помимо этого места, придется собирать модель, что лишь дополнительные расходы

SiMiZZZ commented on src/services/users.py:
```diff
@@ -0,0 +1,23 @@
+async def get_current_user(api_key: str = Header(...)):
+    user = await User.select().where(User.api_key == api_key).first()
+    if not user:
+        raise HTTPException(status_code=401, detail="Неверный API-ключ")
```
не особо нравится raise HTTP ошибки(view слоя) в слое сервиса

SiMiZZZ commented on src/services/auth/login.py:
```diff
@@ -0,0 +1,16 @@
+            status_code=400, detail="Пользователь с таким именем уже существует"
+        )
+    api_key = f"key-{uuid4()}"
+    new_user = User(name=user_data.name, role=user_data.role, api_key=api_key)
```
допустим тут проблема на уровне бд происходит
не дает создать из-за ключа уникальности или типа того
упадем в 500-ку

SiMiZZZ replied:
(надо подумать что может упасть со стороны бд и исправить это выше)
не обязательно отлавливать саму ошибку из бд

SiMiZZZ commented on src/services/auth/login.py:
```diff
@@ -0,0 +1,16 @@
+async def register(user_data: UserCreate):
+    existing_user = await User.select().where(User.name == user_data.name).first()
+    if existing_user:
+        raise HTTPException(
```
также не особо нравится рейз HTTP ошибки

SiMiZZZ commented on src/services/auth/login.py:
```diff
@@ -0,0 +1,16 @@
+
+
+async def register(user_data: UserCreate):
+    existing_user = await User.select().where(User.name == user_data.name).first()
```
тут попадаем в типичное состояние гонки
2 запроса идут в 2 разных транзакциях
1 - проверка наличия пользователя
2 - создание пользователя

представим ситуацию
одновременно отправили 2 запроса на ручку
оба пришли в бд единомоментно и спрашивают - есть ли пользователь? 
бд отвечает что нет(обоим так ответит, потому что его правда еще нет)
и оба побегут дальше создавать пользователя

SiMiZZZ commented on src/app.py:
```diff
@@ -1,7 +1,10 @@
+
 
 @app.get("/")
 async def root() -> dict[str, str]:
```
мертвый ненужный код

AveGus committed:
*Изменена логика на ручке при регистрации
*Полностью изменена бизнес-логика регистрации пользователей
*В Enum пользовательских ролей  enum был изменен на StrEnum
* pydantic модель UserCreate была изменена на NewUser(в соответствие с апи схемой
*В модели CreatedUser был добавлен тайпхинт на роли
*Изменено получение текущего пользователя(get_current_user), согласно апи-схеме

SiMiZZZ commented on src/api/v1/auth.py:
```diff
@@ -0,0 +1,21 @@
+
+
+@router.post("", response_model=CreatedUser)
+async def register_new_user(new_user: NewUser) -> User:
```
-> CreatedUser

SiMiZZZ commented on src/services/auth/login.py:
```diff
@@ -0,0 +1,25 @@
+        await user.save()
+        return user
+    except UniqueViolationError:
+        return UserAlreadyExistsError("Пользователь уже существует")
```
raise вместо return

SiMiZZZ commented on src/api/v1/auth.py:
```diff
@@ -0,0 +1,21 @@
+        return user
+    except UserAlreadyExistsError as e:
+        raise HTTPException(status_code=400, detail=str(e))
+    except Exception as e:
```
500 итак вернется

SiMiZZZ commented on src/services/auth/login.py:
```diff
@@ -0,0 +1,25 @@
+
+async def registration_new_user(name: str, *, role: str = "USER") -> User:
+    try:
+        if await User.select().where(User.name == name).first():
```
nit: exists() вместо select

SiMiZZZ commented on src/services/auth/login.py:
```diff
@@ -0,0 +1,25 @@
+    pass
+
+
+async def registration_new_user(name: str, *, role: str = "USER") -> User:
```
звездочку первый раз вижу

SiMiZZZ commented on src/services/users.py:
```diff
@@ -0,0 +1,33 @@
+async def get_current_user(
+    authorization: Annotated[str, Header(..., alias="Authorization")],
+):
+    if not authorization.startswith("TOKEN "):
```
Вынести переменную постфикса

SiMiZZZ commented on src/services/users.py:
```diff
@@ -0,0 +1,33 @@
+        )
+    api_key = authorization[6:]
+    user = await User.select().where(User.api_key == api_key).first()
+    if not user:
```
nit: if user is not None

AveGus committed:
*Сконфигурирован pytest(было добавлено создание базы, создание таблиц в базе, очищение базы для тестов)
*Добавлены миграции piccolo
*Depency injection были вынесены в соответствующий файл di
*Была добавлена кастомная ошибка(UserAlreadyExists)
*Бизнес-логика регистрации пользователей была изменена(используется get_or_create)
*Были созданы тесты для регистрации(тест регистрации нового пользователя, проверка на дубликат пользователя)
*В types был добавлен новый класс ApiKey(который генерирует апи ключ, а также применяется в пайдентик модели)
